File: templates/backend/fastapi/app/routes/todos.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select

from app.auth import get_current_user_id
from app.database import async_session
from app.models import Todo

router = APIRouter(prefix="/api/todos", tags=["todos"])


class TodoCreate(BaseModel):
    title: str
    description: str | None = None


class TodoUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    completed: bool | None = None


class TodoResponse(BaseModel):
    id: str
    title: str
    description: str | None
    completed: bool
    created_at: str


@router.get("", response_model=list[TodoResponse])
async def list_todos(user_id: str = Depends(get_current_user_id)) -> list[TodoResponse]:
    async with async_session() as session:
        result = await session.execute(
            select(Todo).where(Todo.user_id == uuid.UUID(user_id)).order_by(Todo.created_at.desc())
        )
        todos = result.scalars().all()
        return [
            TodoResponse(
                id=str(t.id),
                title=t.title,
                description=t.description,
                completed=t.completed,
                created_at=t.created_at.isoformat(),
            )
            for t in todos
        ]


@router.post("", response_model=TodoResponse, status_code=status.HTTP_201_CREATED)
async def create_todo(
    body: TodoCreate, user_id: str = Depends(get_current_user_id)
) -> TodoResponse:
    async with async_session() as session:
        todo = Todo(
            id=uuid.uuid4(),
            user_id=uuid.UUID(user_id),
            title=body.title,
            description=body.description,
        )
        session.add(todo)
        await session.commit()
        await session.refresh(todo)
        return TodoResponse(
            id=str(todo.id),
            title=todo.title,
            description=todo.description,
            completed=todo.completed,
            created_at=todo.created_at.isoformat(),
        )


@router.get("/{todo_id}", response_model=TodoResponse)
async def get_todo(todo_id: str, user_id: str = Depends(get_current_user_id)) -> TodoResponse:
    async with async_session() as session:
        result = await session.execute(
            select(Todo).where(Todo.id == uuid.UUID(todo_id), Todo.user_id == uuid.UUID(user_id))
        )
        todo = result.scalar_one_or_none()
        if not todo:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        return TodoResponse(
            id=str(todo.id),
            title=todo.title,
            description=todo.description,
            completed=todo.completed,
            created_at=todo.created_at.isoformat(),
        )


@router.patch("/{todo_id}", response_model=TodoResponse)
async def update_todo(
    todo_id: str, body: TodoUpdate, user_id: str = Depends(get_current_user_id)
) -> TodoResponse:
    async with async_session() as session:
        result = await session.execute(
            select(Todo).where(Todo.id == uuid.UUID(todo_id), Todo.user_id == uuid.UUID(user_id))
        )
        todo = result.scalar_one_or_none()
        if not todo:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        if body.title is not None:
            todo.title = body.title
        if body.description is not None:
            todo.description = body.description
        if body.completed is not None:
            todo.completed = body.completed
        await session.commit()
        await session.refresh(todo)
        return TodoResponse(
            id=str(todo.id),
            title=todo.title,
            description=todo.description,
            completed=todo.completed,
            created_at=todo.created_at.isoformat(),
        )


@router.delete("/{todo_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_todo(todo_id: str, user_id: str = Depends(get_current_user_id)) -> None:
    async with async_session() as session:
        result = await session.execute(
            select(Todo).where(Todo.id == uuid.UUID(todo_id), Todo.user_id == uuid.UUID(user_id))
        )
        todo = result.scalar_one_or_none()
        if not todo:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        await session.delete(todo)
        await session.commit()

File: templates/backend/express/src/routes/todos.ts
import { Router, Response } from "express";
import { Pool } from "pg";
import { v4 as uuidv4 } from "uuid";
import { authenticate, AuthRequest } from "../middleware/auth.js";

export function todosRouter(pool: Pool): Router {
  const router = Router();
  router.use(authenticate);

  router.get("/", async (req: AuthRequest, res: Response) => {
    const result = await pool.query(
      "SELECT id, title, description, completed, created_at FROM todos WHERE user_id = $1 ORDER BY created_at DESC",
      [req.userId],
    );
    res.json(result.rows);
  });

  router.post("/", async (req: AuthRequest, res: Response) => {
    const { title, description } = req.body;
    if (!title) {
      res.status(400).json({ detail: "Title required" });
      return;
    }

    const id = uuidv4();
    const result = await pool.query(
      "INSERT INTO todos (id, user_id, title, description) VALUES ($1, $2, $3, $4) RETURNING id, title, description, completed, created_at",
      [id, req.userId, title, description || null],
    );
    res.status(201).json(result.rows[0]);
  });

  router.get("/:id", async (req: AuthRequest, res: Response) => {
    const result = await pool.query(
      "SELECT id, title, description, completed, created_at FROM todos WHERE id = $1 AND user_id = $2",
      [req.params.id, req.userId],
    );
    if (result.rows.length === 0) {
      res.status(404).json({ detail: "Todo not found" });
      return;
    }
    res.json(result.rows[0]);
  });

  router.patch("/:id", async (req: AuthRequest, res: Response) => {
    const existing = await pool.query(
      "SELECT id, title, description, completed FROM todos WHERE id = $1 AND user_id = $2",
      [req.params.id, req.userId],
    );
    if (existing.rows.length === 0) {
      res.status(404).json({ detail: "Todo not found" });
      return;
    }

    const current = existing.rows[0];
    const title = req.body.title ?? current.title;
    const description = req.body.description ?? current.description;
    const completed = req.body.completed ?? current.completed;

    const result = await pool.query(
      "UPDATE todos SET title = $1, description = $2, completed = $3 WHERE id = $4 AND user_id = $5 RETURNING id, title, description, completed, created_at",
      [title, description, completed, req.params.id, req.userId],
    );
    res.json(result.rows[0]);
  });

  router.delete("/:id", async (req: AuthRequest, res: Response) => {
    const result = await pool.query(
      "DELETE FROM todos WHERE id = $1 AND user_id = $2",
      [req.params.id, req.userId],
    );
    if (result.rowCount === 0) {
      res.status(404).json({ detail: "Todo not found" });
      return;
    }
    res.status(204).send();
  });

  return router;
}
